### PR TITLE
AES accelerator implementation

### PR DESCRIPTION
**Pull request by matrixhead**
This is just a draft version, I'd like to hear everyone's thoughts on this, also i want to implement it for other chips as well.

### CODE REVIEW

**Comment by bjoernQ** [comment]
Already looks good to me! Looking forward to see support for the other chips 👍 Thanks!

**Comment by jessebraham** [comment]
Thanks for opening this PR! I've just converted this PR to a draft until it's ready, mostly just to help me keep track of things 😁

**Comment by matrixhead** [comment]
hey guys i updated the commit could you please take a look at that.  @jessebraham @bjoernQ

**Comment by bjoernQ** [comment]
Great job! Looks good to me!

If you don't mind it would be good to have this rebased on origin/main, then we can enable CI (it won't work otherwise because of the yanked `riscv` crate)

Thanks, again!

**Comment by bjoernQ** [comment]
@jessebraham this looks good to me and worked when testing locally - what do you think?

File: esp-hal-common/src/aes/mod.rs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        generic::{Readable, Reg, RegisterSpec, Resettable, Writable},
+        AES,
+    },
+    system::PeripheralClockControl,
+};
+
+#[cfg_attr(esp32, path = "esp32.rs")]
+#[cfg_attr(esp32s3, path = "esp32_s3.rs")]
+#[cfg_attr(esp32s2, path = "esp32_s2.rs")]
+#[cfg_attr(esp32c3, path = "esp32_c3.rs")]
```
Comment (jessebraham) [inline esp-hal-common/src/aes/mod.rs:33]:
```suggestion
#[cfg_attr(esp32s3, path = "esp32s3.rs")]
#[cfg_attr(esp32s2, path = "esp32s2.rs")]
#[cfg_attr(esp32c3, path = "esp32c3.rs")]
```

Sorry for this, but we do not use underscores in other files like these in the HAL (`esfuse/esp32xx.rs`, `clock/clocks_ll/esp32xx.rs`, etc.), so would you mind please changing these?

**Review by jessebraham — CHANGES_REQUESTED** [review]
Overall this is looking quite good, thanks for the contribution!

I left one small comment, a bit nit-picky but it's in the interest of keeping things consistent.

This seems to be working for the Xtensa chips, however when I tested the ESP32-C3 it simply printed "done" after the second-stage bootloader output, it never printed the expected 4 lines comparing HW/SW AES.

**Comment by bjoernQ** [comment]
> i kind of cheated on that 😬 i couldn't figure out how to get cpu cycles of risc cpus so i just skipped that part, no problem I'll figure it out and make the necessary changes right away 😄

I think it's similar for the SHA example on ESP32-C3 - there we just print the output as hex bytes without a comparison of the CPU cycles - I guess that should be fine

**Comment by matrixhead** [comment]
> > i kind of cheated on that 😬 i couldn't figure out how to get cpu cycles of risc cpus so i just skipped that part, no problem I'll figure it out and make the necessary changes right away 😄
> 
> I think it's similar for the SHA example on ESP32-C3 - there we just print the output as hex bytes without a comparison of the CPU cycles - I guess that should be fine

yes that's what i referred to when I'm writing this, i tried searching for a way to this and I'm not sure about this can i use [mcycle](https://docs.rs/riscv/latest/riscv/register/mcycle/index.html) for this?

**Comment by bjoernQ** [comment]
As far as I know MCYCLE isn't supported - maybe `SystemTimer::now()` works for you?

**Comment by matrixhead** [comment]
> As far as I know MCYCLE isn't supported - maybe `SystemTimer::now()` works for you? (It's not cycles but for comparing SW and HW it should be okay)

okay let me try this.

**Review by jessebraham — APPROVED** [review]
Thanks for making those changes, LGTM!

**Pull request merged**